lua/ttt2/libraries/thermalvision.lua

```diff
@@ -0,0 +1,263 @@
+			render.SetStencilPassOperation(STENCIL_KEEP)
+			render.SetStencilFailOperation(STENCIL_KEEP)
+
+			DrawColorModify(colorModifyEnts)
+		end
+
+		render.SetStencilEnable(false)
+	end
+
+	---
+	-- Hook adding
+	-- @realm client
+	local function AddThermalvisionHook()
+		if thermalvisionHookInstalled then return end
+
+		hook.Add("PostDrawTranslucentRenderables", "RenderThermalvision", RenderHook)
+
+		thermalvisionHookInstalled = true
+	end
+
+	---
+	-- Hook removing
+	-- @realm client
+	local function RemoveThermalvisionHook()
+		hook.Remove("PostDrawTranslucentRenderables", "RenderThermalvision")
+
+		thermalvisionHookInstalled = false
+	end
+
+	---
+	-- Clearing the cached @{Entity} list
+	-- @realm client
+	function thermalvision.Clear()
+		thermalvisionList = {}
+
+		RemoveThermalvisionHook()
+	end
+
+	---
+	-- internal entity removing
+	-- @realm client
+	local function RemoveInternal(ents)
+		local thermalvisionListSize = #thermalvisionList
+		if thermalvisionListSize == 0 or #ents == 0 then return end
+
+		for i = 1, #ents do
+			for j = 1, #thermalvisionList do
+				if thermalvisionList[j].ent == ents[i] then
+					--for now only setting it to nil
+					thermalvisionList[j] = nil
+				end
+			end
+		end
+
+		--cleanup table by 
+		table.RemoveEmptyEntries(thermalvisionList, thermalvisionListSize)
+	end
+
+	---
+	-- Removes entities from the @{Entity} list
+	-- @param table ents list of entities that should get removed
+	-- @realm client
+	function thermalvision.Remove(ents)
+		RemoveInternal(ents)
+
+		if table.IsEmpty(thermalvisionList) and not bgColoring then
+			RemoveThermalvisionHook()
+		end
+	end
+
+	---
+	-- Adds entities into the @{Entity} list that should be rendered with thermalvision
+	-- @param table ents list of @{Entity} that should be added
+	-- @param[default=THERMALVISION_MODE_BOTH] enum mode when should the entity be rendererd
+	-- @realm client
+	function thermalvision.Add(ents, mode)
+		if #ents == 0 then return end
+
+		mode = mode or THERMALVISION_MODE_BOTH
+
+		for i = 1, #ents do
+			thermalvisionList[#thermalvisionList + 1] = {ent = ents[i], mode = mode}
+		end
+
+		-- add the hook if there is something to render
+		AddThermalvisionHook()
+	end
+
+	---
+	-- Enables/disables rendering the background in the thermalvision typical blue tone
+	-- @param bool enabled whether or not the background should get colored
+	-- @realm client
+	function thermalvision.SetBackgroundColoring(enabled)
+		bgColoring = enabled
+	end
+
+	---
+	-- Sets entities of the @{Entity} list to the specified mode
+	-- @param table ents list of @{Entity} that should be set
+	-- @param enum mode when should the entity be rendererd
+	-- @realm client
+	function thermalvision.Set(ents, mode)
+		if #ents == 0 then return end
+
+		-- check if an entity is already inserted and remove it
+		RemoveInternal(ents)
+
+		for i = 1, #ents do
+			thermalvisionList[#thermalvisionList + 1] = {ent = ents[i], mode = mode}
```

Comment:
localizing `#thermalvisionList` outside the loop and then just doing `thermalvisionListSize + i` would be faster